ostd/src/task/scheduler/mod.rs

```diff
@@ -0,0 +1,292 @@
+// SPDX-License-Identifier: MPL-2.0
+
+//! Scheduling subsystem (in-OSTD part).
+//!
+//! This module defines what OSTD expects from a scheduling implementation
+//! and provides useful functions for controlling the execution flow.
+#![allow(dead_code)]
+
+mod fifo_scheduler;
+
+pub use fifo_scheduler::FifoScheduler;
+use spin::Once;
+
+use super::{processor, task::Task, TaskStatus};
+use crate::{
+    arch::{self, timer},
+    cpu::this_cpu,
+    prelude::*,
+};
+
+static SCHEDULER: Once<&'static dyn Scheduler<Task>> = Once::new();
+
+static SHOULD_PREEMPT: ShouldPreemptFlag = ShouldPreemptFlag::new();
+
+/// Injects a scheduler implementation into framework.
+///
+/// This function can only be called once and must be called during the initialization of kernel.
+pub fn inject_scheduler(scheduler: &'static dyn Scheduler<Task>) {
+    SCHEDULER.call_once(|| scheduler);
+
+    timer::register_callback(|| {
+        SCHEDULER.get().unwrap().local_mut_rq_with(&mut |local_rq| {
+            if local_rq.update_current(UpdateFlags::Tick) {
+                SHOULD_PREEMPT.set();
+            }
+        })
+    });
+}
```

Comment:
> However the right way to do this is to call `preempt()` in a timer IRQ handler.

IMO, Switching tasks should be avoided during any kernel interrupt context mostly because the interrupted code point may not be **"interrupt-safe",** which accounts for deadlocks, resource conflicts and other logical race conditions. Besides, it may lower the overall system responsiveness and reduce the readibility and maintainability of the code, which is not adopted in the Linux kernel except for some extreme real-time cases.

The `SHOULD_PREEMPT` flag could be implemented better (at least avoid dumping assembly) though.

Reply:
> No Forced Preemption (server) – the traditional Linux preemption model, geared towards throughput . System call returns and interrupts are the only preemption points.

https://wiki.linuxfoundation.org/realtime/documentation/technical_basics/preemption_models

To my knowledge, the least preemptive preemption model makes schedule points in IRQ contexts. Am I wrong?